arch/arm/core/aarch64/mmu/arm_mmu.c

```diff
@@ -185,14 +95,30 @@ static void set_pte_block_desc(uint64_t *pte, uint64_t addr_pa,
 {
 	uint64_t desc = addr_pa;
 	unsigned int mem_type;
+	unsigned int rwx_el0, rwx_elh;
+
+	rwx_el0 = GET_RWX_EL0(attrs);
+	rwx_elh = GET_RWX_ELh(attrs);
 
 	desc |= (level == 3) ? PTE_PAGE_DESC : PTE_BLOCK_DESC;
 
 	/* NS bit for security memory access from secure state */
 	desc |= (attrs & MT_NS) ? PTE_BLOCK_DESC_NS : 0;
 
-	/* AP bits for Data access permission */
-	desc |= (attrs & MT_RW) ? PTE_BLOCK_DESC_AP_RW : PTE_BLOCK_DESC_AP_RO;
+	/*
+	 * AP bits for EL0 / ELh Data access permission
+	 *
+	 *   AP[2:1]   ELh  EL0
+	 * +--------------------+
+	 *     00      RW   NA
+	 *     01      RW   RW
+	 *     10      RO   NA
+	 *     11      RO   RO
+	 */
+	desc |= (rwx_elh & MT_W) ? PTE_BLOCK_DESC_AP_RW :
+				   PTE_BLOCK_DESC_AP_RO;
+	desc |= (rwx_el0 & MT_RW) ? PTE_BLOCK_DESC_AP_EL0 :
+				    PTE_BLOCK_DESC_AP_ELh;
```

Comment:
1) `PTE_BLOCK_DESC_AP_RW`/`PTE_BLOCK_DESC_AP_RO `(bit 7) has to be set based on RW/RO permission passed from top; it applies even for EL0, checking just for higher EL(`rwx_elh`) seems wrong.
2) `PTE_BLOCK_DESC_AP_EL0`/ `PTE_BLOCK_DESC_AP_ELh` (bit 6) has to be set for EL0 access, it has nothing to do with MT_RW, meaning it applies even for MT_RO.

So to sort out this confusion, bit 7 has to be set for RO/RW and for access to EL0, bit 6 has to be set. - it is as simple as that. It can't be like ELh has only RO permission and EL0 has RW permission. Description given below for AP[2:1] field also clarifies the same.

Reply:
Let's start from line 120-121:
- I'm checking if in `rwx_el0` either `R` or `W` is set. I'm basically checking if I have set any `RW` permission for EL0. If one of the two bits is set, then I'm setting the bit `AP[1]` to `1`. This is basically line 2 and 4 in the table. Please note that this is verified also for `RO` since `MT_RW` is an OR between `W` and `R`, so it does apply even for `MT_RO`.

Line 118-119:
- If you look at the `AP` table lines 1 and 2 you can see that we are setting `AP[2]` to 0 every time we have `W`, and on the contrary if we do not have `W` we are setting that bit to 1. This is why I'm only checking for `W` in this case.